hwmgr-plugins/metal3/controller/helpers.go

```diff
@@ -1191,3 +1209,137 @@ func validateNodeConfiguration(
 	logger.InfoContext(ctx, "Firmware versions and BIOS settings validated successfully", slog.String("BMH", bmh.Name))
 	return true, nil
 }
+
+// propagateTimeoutToAllocatedNodes propagates timeout state from NodeAllocationRequest to all associated AllocatedNodes
+func propagateTimeoutToAllocatedNodes(
+	ctx context.Context,
+	c client.Client,
+	noncachedClient client.Reader,
+	logger *slog.Logger,
+	nodeAllocationRequest *pluginsv1alpha1.NodeAllocationRequest,
+	conditionType hwmgmtv1alpha1.ConditionType,
+	timeoutMessage string,
+) error {
+	logger.InfoContext(ctx, "Propagating timeout state to AllocatedNodes",
+		slog.String("nodeAllocationRequest", nodeAllocationRequest.Name),
+		slog.String("conditionType", string(conditionType)))
+
+	// Get all AllocatedNodes associated with this NodeAllocationRequest
+	nodeList, err := hwmgrutils.GetChildNodes(ctx, logger, c, nodeAllocationRequest)
+	if err != nil {
+		return fmt.Errorf("failed to get AllocatedNodes for NodeAllocationRequest %s: %w", nodeAllocationRequest.Name, err)
+	}
+
+	if len(nodeList.Items) == 0 {
+		logger.InfoContext(ctx, "No AllocatedNodes found for NodeAllocationRequest",
+			slog.String("nodeAllocationRequest", nodeAllocationRequest.Name))
+		return nil
+	}
+
+	// Update each AllocatedNode with timeout state
+	// Use the timeout message directly without prepending NAR name to avoid message nesting
+	for _, node := range nodeList.Items {
+		if err := hwmgrutils.SetNodeConditionStatus(
+			ctx, c, noncachedClient,
+			node.Name, node.Namespace,
+			string(conditionType),
+			metav1.ConditionFalse,
+			string(hwmgmtv1alpha1.TimedOut),
+			timeoutMessage, // Use message directly without prepending NAR name
+		); err != nil {
+			logger.ErrorContext(ctx, "Failed to propagate timeout to AllocatedNode",
+				slog.String("nodeAllocationRequest", nodeAllocationRequest.Name),
+				slog.String("allocatedNode", node.Name),
+				slog.String("error", err.Error()))
+			// Continue with other nodes even if one fails
+		} else {
+			logger.InfoContext(ctx, "Successfully propagated timeout to AllocatedNode",
+				slog.String("nodeAllocationRequest", nodeAllocationRequest.Name),
+				slog.String("allocatedNode", node.Name))
+		}
+	}
+
+	logger.InfoContext(ctx, "Completed timeout propagation to AllocatedNodes",
+		slog.String("nodeAllocationRequest", nodeAllocationRequest.Name),
+		slog.Int("nodeCount", len(nodeList.Items)))
+	return nil
+}
+
+// clearBMHUpdateAnnotationsForNAR removes BIOS and firmware update annotations from all BMHs
```

Comment:
Why do we need these annotations ?

Reply:
The current implementation updates the HFC and monitors BMH for completion. The UpdateNeededAnnotation is used as a coordination mechanism, not a replacement for HFC.